src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -411,6 +447,149 @@ void ReduceAxesCompute(const nnvm::NodeAttrs& attrs,
   ReduceAxesComputeImpl<xpu, reducer, normalize>(attrs, ctx, inputs, req, outputs, small);
 }
 
+template <int req, int axis>
+struct SumCsrKernel;
+
+template <int req>
+struct SumCsrKernel<req, 0> {
+  template <typename RType, typename IType, typename DType>
+  MSHADOW_XINLINE static void Map(int j, DType* out_data,
+                                  const RType* in_indptr, const IType* in_idx,
+                                  const DType* in_data,
+                                  const int64_t num_rows) {
+    DType sum, residual;
```

Comment:
I think each thread should handle multiple/a range of output elements instead of just one so that you perform fewer binary searches. We can use the temp resource to store temp sum result. We can use num_cpu_thread to invoke the kernel instead of size_of_output. Also if nnz per row is fewer than 16 then I guess linear search instead of binary search would be faster..

Reply:
You can look at cast_storage gpu implementation to see how to request a temp resource

Reply:
It would be more consistent to have num_rows as RType. Also will avoid signed/unsigned problems when RType varies in the future.

Reply:
@eric-haibin-lin  Great suggestion ! I have changed the logic to handle a range instead of individual element.  @cjolivier01 Good catch!  I am using RType and IType for num_rows now.